### ISSUE TITLE
pad function alignment with ud2

### ISSUE THREAD

**Issue by nickdesaulniers** [ClangBuiltLinux/linux#976]
We had a bug in asm goto w/ output support that Linus reported; it's since been fixed, but we were falling through one function into the next.  I also get 0day reports from objtool about falling through to the next functin (TODO: Nick: file bug).  Linus had the idea to use `ud2` (or equivalent undefined instruction) for the padding between functions, rather than `.p2align` which expands to nops.

`ud2` on x86 is encoded in 2B, so it can't do 1B realignment.  I assume for cases where we'd need to realign functions, there will be waste from this feature.

This is a great idea for a general security feature, one that would make sense beyond even the kernel I think.  cc @kees @ramosian-glider 

@kbeyls does ARM have a dedicated assembler mnemonic for "trigger an undefined instruction exception?"  I assume it's always going to be 2B (thumb) / 4B (arm), so we'd always be inserting that instruction, followed by a realign w/ nops?

**Comment by kees** [comment]
For 1 byte alignment, I recommend 0xCC (int 3) -- this is what I'm using for section padding in the x86 linker script:
https://git.kernel.org/linus/7705dc8557973d8ad8f10840f61d8ec805695e9e

The FG-KASLR series is doing similar:
https://github.com/kaccardi/linux/issues/20
https://github.com/kaccardi/linux/commit/afe11296a2b97c04fd74da88f0b691e12ecd2eec

For arm64, I think you want a "brk" with maybe BUG_BRK_IMM:
see arch/arm64/include/asm/asm-bug.h and arch/arm64/include/asm/brk-imm.h
(I can't define 4-byte padding patterns in linker scripts so I can't do similar to x86's linker script.)

**Comment by nickdesaulniers** [comment]
It would nice for this to be an `-f<foo>` flag to the compiler so that non-kernel projects could have the same protections without needing to write linker scripts.  I never even knew about linker scripts until working on the kernel (and now I know about them than I care to).

**Comment by ihalip** [comment]
ARM has `udf`, which is 4 bytes (2 bytes with -mthumb).

I think in general you'd want the padding instruction to be whatever `__builtin_trap()` would generate (or a smaller, close equivalent).

**Comment by kbeyls** [comment]
> @kbeyls does ARM have a dedicated assembler mnemonic for "trigger an undefined instruction exception?" I assume it's always going to be 2B (thumb) / 4B (arm), so we'd always be inserting that instruction, followed by a realign w/ nops?

As @ihalip indicated:
For AArch64, UDF #<imm> exists. As all other AArch64 instructions, that's a 4-byte encoding.
For AArch32, UDF<c> #<imm8>, UDF<c>.W #<imm16> and UDF<c> #<imm16> exists. The first one is a 2-byte encoding, the others are 4-byte encodings.

Of course, maybe some of the alternatives, such as BRK pointed out by @kees might be a more appropriate use in the kernel (just guessing here).

**Comment by kbeyls** [comment]
If I understood the suggestion here correctly, isn't it really about adding a new option to the .p2align assembler directive to fill using instructions that will stop the execution, rather than with NOPs?
And then maybe make compilers use that when they can prove that the instructions emitted by .p2align should never be executed?

**Comment by nickdesaulniers** [comment]
In [this thread](https://bugs.llvm.org/show_bug.cgi?id=50080#c5) @topperc mentions that `-mllvm -trap-unreachable` can at least add traps to unreachable blocks.  I'm not sure that will additionally protect functions without unreachable blocks, but perhaps worth evaluating.

**Comment by nickdesaulniers** [comment]
@MaskRay do you know of any existing functionality the linker may provide to pad functions with specific instructions? I think we can specify padding between sections with explicit byte patterns via linker scripts, but not functions within the same section AFAIK.

**Comment by MaskRay** [comment]
I concur with https://isopenbsdsecu.re/mitigations/trapsled/

> But on the other hand, this is a super-cheap change, that doesn’t introduce complexity at all, and might help to catch some misalignment bugs, but this definitively isn’t, by any means, a security feature.

There is an output section fill feature in GNU ld and LLD but care is needed for instructions longer than 1 on x86 when the output section may end at an instruction boundary.

```
% grep 0xcccccccc arch/x86/kernel/vmlinux.lds.S
        } :text = 0xcccccccc
```

See https://maskray.me/blog/2020-12-19-lld-and-gnu-linker-incompatibilities#output-section-fill that the pattern needs to be a 4-byte word.

For a filler for the align directive in assemblers, we have https://sourceware.org/binutils/docs/as/Align.html

**Comment by nickdesaulniers** [comment]
> There is an output section fill feature in GNU ld and LLD but care is needed for instructions longer than 1 on x86 when the output section may end at an instruction boundary.

Right, @kees implemented that for x86 in https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=7705dc8557973d8ad8f10840f61d8ec805695e9e.

That's a good first start; section padding is a large granularity. Function padding would be a finer granularity.
